## 문제107. 이미지 이름앞에 절대 경로가 아래처럼 붙게 하시오

In [ ]:
import os 
import re
train_image = 'D:\\data\\cifar10\\train100'
    
def image_load(path):
    file_list = os.listdir(path)
    file_name = []
    for i in file_list:
        a = int(re.sub('[^0-9]', '', i)) #숫자가 아닌것은 ''으로 처리
        file_name.append(a)
        file_name.sort()
    file_res = []
    for j in file_name:
        file_res.append('%s\\%d.png' %(path,j))
    
    return file_res

print(image_load(train_image))

## 문제108. cv2.imread 함수를 이용해서 이미지를 숫자로 변경하시오

아나콘다 프롬프트창에서

activate snowdeer_env  
conda install opencv  

In [7]:
import  os
import  re
import  cv2
import  numpy  as  np
train_image = 'D:\\data\\cifar10\\train100'

def  image_load(path):
    file_list = os.listdir(path)
    file_name = []
    for  i  in  file_list:
        a = int(  re.sub('[^0-9]', '', i )  ) # 숫자가 아닌것은 '' 로 처리 
        file_name.append(a) 
    file_name.sort()

    file_res = [] 
    for  j  in   file_name:
        file_res.append('%s\\%d.png' %(path,j)  )

    image = []
    for  k  in  file_res:
        img = cv2.imread(k)
        image.append(img)

    return np.array(image)
   
print ( image_load( train_image) )

[[[ 44  64  62]
  [ 26  50  50]
  [ 19  44  46]
  ...
  [ 69 172 167]
  [ 76 184 183]
  [ 72 136 137]]

 [[ 37  65  63]
  [ 26  53  55]
  [ 27  50  52]
  ...
  [ 61 169 163]
  [ 75 174 171]
  [ 77 146 145]]

 [[ 36  62  58]
  [ 37  66  64]
  [ 37  60  56]
  ...
  [ 62 155 153]
  [ 64 154 150]
  [ 57 128 123]]

 ...

 [[ 99 135 172]
  [ 84 110 143]
  [ 42  56 130]
  ...
  [ 56  75  94]
  [ 86 108 141]
  [ 81 105 139]]

 [[117 146 183]
  [ 95 118 150]
  [ 44  64  80]
  ...
  [ 60  72  81]
  [ 98 118 135]
  [110 125 143]]

 [[144 174 209]
  [123 151 182]
  [ 83 109 139]
  ...
  [ 47  54  59]
  [111 119 130]
  [160 156 169]]]


## 문제109. 훈련이미지의 라벨을 one hot encoding 하는 코드 구현 하시오 !

In [ ]:
train_label = 'D:\\data\\cifar10\\train_label.csv'
import   csv

def  label_load( path ):
    file = open(path)
    labeldata = csv.reader(file)
    labellist = []
    for  i   in  labeldata:
        labellist.append(i)
    return  labellist 

print ( label_load(train_label) )

## 문제110. 위의 결과가 문자가 아니라 숫자로 출력되게 하시오!

In [8]:
train_label = 'D:\\data\\cifar10\\train_label.csv'
import   csv

def  label_load( path ):
    file = open(path)
    labeldata = csv.reader(file)
    labellist = []
    for  i   in  labeldata:
        labellist.append(i)

    label = np.array(labellist)
    label = label.astype(int)  # 숫자로 변환 
    return  label

print ( label_load(train_label) )

[[6]
 [9]
 [9]
 ...
 [9]
 [1]
 [1]]


## 문제111. 아래의 결과를 출력하시오!

[0 0 0 0 1 0 0 0 0 0 0 ]

In [6]:
import numpy as np

print(np.eye(10)[4])

[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


## 문제112. 위의 np.eye를 가지고 문제 10번에서 출력된 숫자들이 아래와 같이 one hot encoding 된 숫자로 출력되게하시오.


[ 0 0 0 0 1 0 0 0 0 0 ]  
[ 0 0 1 0 0 0 0 0 0 0 ]  
[ 0 0 0 0 0 1 0 0 0 0 ]  
..

In [10]:
import csv
def label_load(path):
    file = open(path)
    labeldata = csv.reader(file)
    labellist = []
    for i in labeldata:
        labellist.append(i)
    label = np.array(labellist)
    label = label.astype(int) # 숫자로 변환
    label = np.eye(10)[label] # 열이 10개인 np.array를 만들어서 label의 값을 index로 가져서 그 위치를 1로 바꿈.
    return label

train_label = 'D:\\data\\cifar10\\train_label.csv'
print(label_load(train_label))

[[[0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 1.]]

 [[0. 0. 0. ... 0. 0. 1.]]

 ...

 [[0. 0. 0. ... 0. 0. 1.]]

 [[0. 1. 0. ... 0. 0. 0.]]

 [[0. 1. 0. ... 0. 0. 0.]]]


## 문제113. 위의 결과는 3차원인데 신경망에서 라벨을 사용하려면 2차원이어야 하므로 2차원으로 변경하시오!

In [ ]:
y = tf.placeholder('float', [None, 10]) # one hot encoding 된 행렬을 담는 변수

# (60000, 1, 10) <--- 60000만장 1행 10열 

#(60000,10) <--- 신경망에 넣기위해서 이렇게 2차원으로 만들어 주어야됨.

In [15]:
import numpy as np

y = np.random.randn(60000,1,10)
y2 = y.reshape(-1,10)
print(y2.shape)

(60000, 10)


## 문제114. 2차원 데이터로 출력되겠금 label_load 함수를 수정하시오.

In [ ]:
import numpy as np
import csv

def label_load(path):
    file = open(path)
    labeldata = csv.reader(file)
    labellist = []
    for i in labeldata:
        labellist.append(i)
    label = np.array(labellist)
    label = label.astype(int) # 숫자로 변환
    label = np.eye(10)[label] # 열이 10개인 np.array를 만들어서 label의 값을 index로 가져서 그 위치를 1로 바꿈.
    label = label.reshape(-1,10)
    return label

train_label = 'D:\\data\\cifar10\\train_label.csv'
print(label_load(train_label))

## ■ cifar10 데이터를 신경망에 로드하기 위해 만든 함수 2가지

1. image_load
2. label_load

## 문제115. 지금까지 만든 두개의 함수 image_load와 label_load를 loader2.py 라는 파이썬 코드에 저장하고 아래와 같이 loader2.py 를 import 한 후에 cifar10 전체 데이터를 로드하는 코드를 구현하시오!

In [ ]:
import loader2

train_image = 'D:\\data\\cifar10\\train\\'
train_label = 'D:\\data\\cifar10\\train_label.csv'

test_image = 'D:\\data\\cifar10\\test\\'
test_label = 'D:\\data\\cifar10\\test_label.csv'

trainX = loader2.image_load(train_image)
trainY = loader2.label_load(train_label)
testX = loader2.image_load(test_image)
testY = loader2.label_load(test_label)

print(trainX.shape)
print(trainY.shape)
print(testX.shape)
print(testY.shape)

"""
(50000, 32, 32, 3)
(50000, 10)
(10000, 32, 32, 3)
(10000, 10)

"""

## ■ cifar10 데이터를 신경망에 로드하기 위해 만든 함수 2가지

1. image_load
2. label_load
3. next_batch

## 문제116. 신경망에 100장씩 데이터를 로드할 수 있도록 아래와 같이 next_batch 함수를 생성하시오!

In [ ]:
def next_batch( data1, data2, init, fina ):
    return data1[ init : fina ], data2[ init : fina ]

test_image = 'D:\\data\\cifar10\\test\\'
test_label = "D:\\data\\cifar10\\test_label.csv"

testX = loader2.image_load(test_image)
testY = loader2.label_load(test_label)

print(next_batch( testX, testY, 0, 100))

## 문제117. 아래와 같이 shuffle_batch 함수를 만들고 loader2.py에 추가 시키시오.

In [ ]:
def shuffle_batch(data_list, label):
    x = np.arange(len(data_list))
    random.shuffle(x)
    data_list2 = data_list[x]
    label2 = label[x]
    
    return data_list2, label2

In [ ]:
# 실행방법:
import loader2

test_image = 'D:\\data\\cifar10\\test\\'
test_label = 'D:\\data\\cifar10\\test_label.csv'

testX = loader2.image_load(test_image)
testY = loader2.label_load(test_label)

print(loader2.shuffle_batch(testX, testY))

## ■ cifar10 데이터를 로드하는 4개의 함수를 이용해서 vgg6 신경망으로 학습하는 전체 코드를 실행하시오 !

In [ ]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import loader2

train_image = 'D:\\data\\cifar10\\train\\'
train_label = 'D:\\data\\cifar10\\train_label.csv'
test_image = 'D:\\data\\cifar10\\test\\'
test_label = 'D:\\data\\cifar10\\test_label.csv'

print("LOADING DATA")

trainX = loader2.image_load(train_image)
trainY = loader2.label_load(train_label)
testX = loader2.image_load(test_image)
testY = loader2.label_load(test_label)

print("LOADED DATA")

tf.reset_default_graph()

#입력층
x = tf.placeholder("float",[None,32,32,3])
keep_prob = tf.placeholder("float")


#conv_1
b1 = tf.Variable(tf.ones([128]))
W1 = tf.Variable(tf.random_normal([3,3,3,128],stddev = 0.01))
y1 = tf.nn.conv2d(x, W1, strides=[1,1,1,1], padding = 'SAME')
y1 = y1 + b1
y1 = tf.contrib.layers.batch_norm(y1,scale=True)
y1 = tf.nn.leaky_relu(y1)

#conv_2
b1_2 = tf.Variable(tf.ones([128]))
W1_2 =  tf.Variable(tf.random_normal([3,3,128,128],stddev = 0.01))
y1_2 = tf.nn.conv2d(y1, W1_2, strides=[1,1,1,1], padding = 'SAME')
y1_2 = y1_2 + b1_2
y1_2 = tf.contrib.layers.batch_norm(y1_2,scale=True)
y1_2 = tf.nn.leaky_relu(y1_2)
y1_2 = tf.nn.dropout(y1_2, keep_prob)

#maxpooling
y1_2 = tf.nn.max_pool(y1_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#conv_3
b2 = tf.Variable(tf.ones([256]))
W2 = tf.Variable(tf.random_normal([3,3,128,256],stddev = 0.01))
y2 = tf.nn.conv2d(y1_2, W2, strides=[1,1,1,1], padding = 'SAME')
y2 = y2 + b2
y2 = tf.contrib.layers.batch_norm(y2,scale=True)
y2 = tf.nn.leaky_relu(y2)

#conv_4
b2_2 = tf.Variable(tf.ones([256]))
W2_2 = tf.Variable(tf.random_normal([3,3,256,256],stddev = 0.01))
y2_2 = tf.nn.conv2d(y2, W2_2, strides=[1,1,1,1], padding = 'SAME')
y2_2 = y2_2 + b2_2
y2_2 = tf.contrib.layers.batch_norm(y2_2,scale=True)
y2_2 = tf.nn.leaky_relu(y2_2)
y2_2 = tf.nn.dropout(y2_2, keep_prob)

#maxpooling
y2_2 = tf.nn.max_pool(y2_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#conv_5
b3 = tf.Variable(tf.ones([512]))
W3 = tf.Variable(tf.random_normal([3,3,256,512],stddev = 0.01))
y3 = tf.nn.conv2d(y2_2, W3, strides=[1,1,1,1], padding = 'SAME')
y3 = y3 + b3
y3 = tf.contrib.layers.batch_norm(y3,scale=True)
y3 = tf.nn.leaky_relu(y3)

#conv_6
b3_2 = tf.Variable(tf.ones([512]))
W3_2 = tf.Variable(tf.random_normal([3,3,512,512],stddev = 0.01))
y3_2 = tf.nn.conv2d(y3, W3_2, strides=[1,1,1,1], padding = 'SAME')
y3_2 = y3_2 + b3_2
y3_2 = tf.contrib.layers.batch_norm(y3_2,scale=True)
y3_2 = tf.nn.leaky_relu(y3_2)
y3_2 = tf.nn.dropout(y3_2, keep_prob)

#maxpooling
y3_2 = tf.nn.max_pool(y3_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#Affine1
b4 = tf.Variable(tf.ones([1024]))
W4 = tf.get_variable(name='W4', shape=[4*4*512, 1024], initializer=tf.contrib.layers.variance_scaling_initializer())
y4 = tf.reshape(y3_2, [-1, 4*4*512])
y4 = tf.matmul(y4,W4) + b4
y4 = tf.contrib.layers.batch_norm(y4,scale=True)
y4 = tf.nn.leaky_relu(y4)


#Affine2
b5 = tf.Variable(tf.ones([1024]))
W5 = tf.get_variable(name='W5', shape=[1024, 1024], initializer=tf.contrib.layers.variance_scaling_initializer())
y5 = tf.matmul(y4,W5) + b5
y5 = tf.contrib.layers.batch_norm(y5,scale=True)
y5 = tf.nn.leaky_relu(y5)


#드롭아웃
y5_drop = tf.nn.dropout(y5, keep_prob)


#출력층
b6 = tf.Variable(tf.ones([10]))
W6 = tf.get_variable(name='W6', shape=[1024, 10], initializer=tf.contrib.layers.variance_scaling_initializer()) # he 초기값
y6 = tf.matmul(y5_drop,W6) + b6
y6 = tf.contrib.layers.batch_norm(y6,scale=True)
y_hat = tf.nn.softmax(y6)


#예측값
y_predict = tf.argmax(y_hat,1)


# 라벨을 저장하기 위한 변수 생성
y_onehot = tf.placeholder("float",[None,10])
y_label = tf.argmax(y_onehot, axis = 1)


# 정확도를 출력하기 위한 변수 생성
correct_prediction = tf.equal(y_predict, y_label)
accuracy = tf.reduce_mean(tf.cast(correct_prediction,"float"))


# 교차 엔트로피 오차 함수
loss = -tf.reduce_sum(y_onehot * tf.log(y_hat), axis = 1)


# SGD 경사 감소법
# optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.05)


# Adam 경사 감소법
optimizer = tf.train.AdamOptimizer(learning_rate=0.001)


# 학습 오퍼레이션 정의
train = optimizer.minimize(loss)


# 변수 초기화
init = tf.global_variables_initializer()
train_acc_list = []
test_acc_list = []


with tf.Session() as sess:
    sess.run(init)
    for j in range(20):
        for i in range(600):
            trainX , trainY = loader2.shuffle_batch(trainX, trainY)
            testX, testY = loader2.shuffle_batch(testX, testY)

            train_xs, train_ys = loader2.next_batch(trainX, trainY, 0, 100)
            test_xs, test_ys = loader2.next_batch(testX, testY, 0, 100)

            sess.run(train, feed_dict={x: train_xs, y_onehot: train_ys, keep_prob: 0.8})

            if i == 0:
                train_acc = sess.run(accuracy, feed_dict={x: train_xs, y_onehot: train_ys, keep_prob: 1.0})
                test_acc = sess.run(accuracy, feed_dict={x: test_xs, y_onehot: test_ys, keep_prob: 1.0})

                train_acc_list.append(train_acc)
                test_acc_list.append(test_acc)

## ■ gpu 피씨에서 tensor_gpu 설치

1. 먼저 기존에 설치된 tensorflow를 삭제하시오

__아나콘다 프롬프트 창을 열고__

conda remove tensorflow

2. tensorflow-gpy를 설치하시오

conda install tensorflow-gpu

3. cuda를 설치하시오 ~

https://replaydeveloper.tistory.com/entry/Tensorflow-14-개발-환경-설치Windows-10-CUDA-80-cuDNN-v60-GPU-버전

## ■ 현재 코드는 배치정규화가 제대로 안되는 문제가 있으므로 수정하시오!

In [ ]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import loader2

train_image = 'D:\\data\\cifar10\\train\\'
train_label = 'D:\\data\\cifar10\\train_label.csv'
test_image = 'D:\\data\\cifar10\\test\\'
test_label = 'D:\\data\\cifar10\\test_label.csv'

print("LOADING DATA")

trainX = loader2.image_load(train_image)
trainY = loader2.label_load(train_label)
testX = loader2.image_load(test_image)
testY = loader2.label_load(test_label)

print("LOADED DATA")

tf.reset_default_graph()

#입력층
x = tf.placeholder("float",[None,32,32,3])
keep_prob = tf.placeholder("float")
training = tf.placeholder(tf.bool, name = 'training') 
# 배치정규화가 훈련할때는 True, Test할때는 False로 설정하는 코드
# True 아니면 False가 들어가는 코드

#conv_1
b1 = tf.Variable(tf.ones([128]))
W1 = tf.Variable(tf.random_normal([3,3,3,128],stddev = 0.01))
y1 = tf.nn.conv2d(x, W1, strides=[1,1,1,1], padding = 'SAME')
y1 = y1 + b1
y1 = tf.contrib.layers.batch_norm(y1,scale=True, is_training = training)
y1 = tf.nn.leaky_relu(y1)

#conv_2
b1_2 = tf.Variable(tf.ones([128]))
W1_2 =  tf.Variable(tf.random_normal([3,3,128,128],stddev = 0.01))
y1_2 = tf.nn.conv2d(y1, W1_2, strides=[1,1,1,1], padding = 'SAME')
y1_2 = y1_2 + b1_2
y1_2 = tf.contrib.layers.batch_norm(y1_2,scale=True, is_training = training)
y1_2 = tf.nn.leaky_relu(y1_2)
y1_2 = tf.nn.dropout(y1_2, keep_prob)

#maxpooling
y1_2 = tf.nn.max_pool(y1_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#conv_3
b2 = tf.Variable(tf.ones([256]))
W2 = tf.Variable(tf.random_normal([3,3,128,256],stddev = 0.01))
y2 = tf.nn.conv2d(y1_2, W2, strides=[1,1,1,1], padding = 'SAME')
y2 = y2 + b2
y2 = tf.contrib.layers.batch_norm(y2,scale=True, is_training = training)
y2 = tf.nn.leaky_relu(y2)

#conv_4
b2_2 = tf.Variable(tf.ones([256]))
W2_2 = tf.Variable(tf.random_normal([3,3,256,256],stddev = 0.01))
y2_2 = tf.nn.conv2d(y2, W2_2, strides=[1,1,1,1], padding = 'SAME')
y2_2 = y2_2 + b2_2
y2_2 = tf.contrib.layers.batch_norm(y2_2,scale=True, is_training = training)
y2_2 = tf.nn.leaky_relu(y2_2)
y2_2 = tf.nn.dropout(y2_2, keep_prob)

#maxpooling
y2_2 = tf.nn.max_pool(y2_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#conv_5
b3 = tf.Variable(tf.ones([512]))
W3 = tf.Variable(tf.random_normal([3,3,256,512],stddev = 0.01))
y3 = tf.nn.conv2d(y2_2, W3, strides=[1,1,1,1], padding = 'SAME')
y3 = y3 + b3
y3 = tf.contrib.layers.batch_norm(y3,scale=True, is_training = training)
y3 = tf.nn.leaky_relu(y3)

#conv_6
b3_2 = tf.Variable(tf.ones([512]))
W3_2 = tf.Variable(tf.random_normal([3,3,512,512],stddev = 0.01))
y3_2 = tf.nn.conv2d(y3, W3_2, strides=[1,1,1,1], padding = 'SAME')
y3_2 = y3_2 + b3_2
y3_2 = tf.contrib.layers.batch_norm(y3_2,scale=True, is_training = training)
y3_2 = tf.nn.leaky_relu(y3_2)
y3_2 = tf.nn.dropout(y3_2, keep_prob)

#maxpooling
y3_2 = tf.nn.max_pool(y3_2, ksize = [1,2,2,1], strides = [1,2,2,1], padding = 'SAME')


#Affine1
b4 = tf.Variable(tf.ones([1024]))
W4 = tf.get_variable(name='W4', shape=[4*4*512, 1024], initializer=tf.contrib.layers.variance_scaling_initializer())
y4 = tf.reshape(y3_2, [-1, 4*4*512])
y4 = tf.matmul(y4,W4) + b4
y4 = tf.contrib.layers.batch_norm(y4,scale=True, is_training = training)
y4 = tf.nn.leaky_relu(y4)


#Affine2
b5 = tf.Variable(tf.ones([1024]))
W5 = tf.get_variable(name='W5', shape=[1024, 1024], initializer=tf.contrib.layers.variance_scaling_initializer())
y5 = tf.matmul(y4,W5) + b5
y5 = tf.contrib.layers.batch_norm(y5,scale=True, is_training = training)
y5 = tf.nn.leaky_relu(y5)


#드롭아웃
y5_drop = tf.nn.dropout(y5, keep_prob)


#출력층
b6 = tf.Variable(tf.ones([10]))
W6 = tf.get_variable(name='W6', shape=[1024, 10], initializer=tf.contrib.layers.variance_scaling_initializer()) # he 초기값
y6 = tf.matmul(y5_drop,W6) + b6
y6 = tf.contrib.layers.batch_norm(y6,scale=True, is_training = training)
y_hat = tf.nn.softmax(y6)


#예측값
y_predict = tf.argmax(y_hat,1)


# 라벨을 저장하기 위한 변수 생성
y_onehot = tf.placeholder("float",[None,10])
y_label = tf.argmax(y_onehot, axis = 1)


# 정확도를 출력하기 위한 변수 생성
correct_prediction = tf.equal(y_predict, y_label)
accuracy = tf.reduce_mean(tf.cast(correct_prediction,"float"))


# 교차 엔트로피 오차 함수
loss = -tf.reduce_sum(y_onehot * tf.log(y_hat), axis = 1)


# SGD 경사 감소법
# optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.05)


# Adam 경사 감소법
optimizer = tf.train.AdamOptimizer(learning_rate=0.001)

update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
with tf.control_dependencies(update_ops):
    # Ensures that we execute the update_ops before performing the train_step
    train = tf.train.AdamOptimizer(learning_rate=0.001).minimize(loss)
    
# 학습 오퍼레이션 정의
train = optimizer.minimize(loss)


# 변수 초기화
init = tf.global_variables_initializer()
train_acc_list = []
test_acc_list = []


with tf.Session() as sess:
    sess.run(init)
    for j in range(20):
        for i in range(600):
            trainX , trainY = loader2.shuffle_batch(trainX, trainY)
            testX, testY = loader2.shuffle_batch(testX, testY)

            train_xs, train_ys = loader2.next_batch(trainX, trainY, 0, 100)
            test_xs, test_ys = loader2.next_batch(testX, testY, 0, 100)

            sess.run(train, feed_dict={x: train_xs, y_onehot: train_ys, keep_prob: 0.8, training: True})

            if i == 0:
                train_acc = sess.run(accuracy, feed_dict={x: train_xs, y_onehot: train_ys, keep_prob: 1.0, training : True})
                test_acc = sess.run(accuracy, feed_dict={x: test_xs, y_onehot: test_ys, keep_prob: 1.0, training : False})
                # 훈련되어진 베타와 감마가 test에 사용되려면 training을 false로 설정해야된다.

                train_acc_list.append(train_acc)
                test_acc_list.append(test_acc)

## ※ 배치 정규화 할때 주의할사항

아래의 코드를 추가해야 훈련할 때 만든 배치정규화의 감마와 베타값을 테스트할 때 사용할 수 있게 된다.

https://neurowhai.tistory.com/119

In [ ]:
"""
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)  
with tf.control_dependencies(update_ops):  
    # Ensures that we execute the update_ops before performing the train_step  
    train = tf.train.AdamOptimizer(learning_rate=0.001).minimize(loss)  
"""